## Week 8 Independent Lab: Data Manipulation

Author: Aerionna Skrutvold  
Course: BGEN632 Grad. Intro to Python  

This notebook contains code for the  practice your new skills in data manipulation with the *piping* expression for `pandas`. The program primarily makes use of pandas DataFrames and the  `gh_turnover_four_projects.csv` file contained within the data folder. 

## Import Modules 

In [1]:
import os
import pandas as pd
import numpy as np
from sklearn.model_selection import KFold

## Set Working Directory

In [2]:
os.chdir("C:/Users/gcassist/Desktop/MSBA/Spring 25/Python/week8labs/data")

## Access Data

In [3]:
gh_turnover = pd.read_csv("gh_turnover_four_projects.csv") #open the file 

In [4]:
gh_turnover.info()  # quick inspect

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 473 entries, 0 to 472
Data columns (total 16 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   project_id     473 non-null    int64  
 1   domain         473 non-null    object 
 2   language       473 non-null    object 
 3   windows        473 non-null    int64  
 4   window_idx     473 non-null    int64  
 5   num_team       473 non-null    int64  
 6   num_commits    473 non-null    int64  
 7   blau_gender    473 non-null    float64
 8   Gini_gh_ten    473 non-null    float64
 9   user_id        473 non-null    int64  
 10  Core1          473 non-null    int64  
 11  commits        473 non-null    int64  
 12  propCommits    473 non-null    float64
 13  gender         449 non-null    object 
 14  github_tenure  473 non-null    int64  
 15  leavesNextQ    473 non-null    int64  
dtypes: float64(3), int64(10), object(3)
memory usage: 59.3+ KB


## Rename Columns
* `domain` change to `project_domain`
* `language` change to `project_language`
* `windows` change to `project_age`
* `window_idx` change to `quarter`
* `num_team` change to `team_size`
* `num_commits` change to `project_commits`
* `blau_gender` change to `gender_blau`
* `Gini_gh_ten` change to `gh_tenure_gini`
* `Core1` change to `core_dev`
* `commits` change to `user_commits`
* `propCommits` change to `commit_proportion`
* `github_tenure` change to `gh_tenure`
* `leavesNextQ` change to `leaves_next_q`
* Leave all other column names as is

In [5]:
(gh_turnover
 .rename({'domain':'project_domain',
          'language':'project_language',
          'windows' : 'project_age',
          'window_idx' : 'quarter',
          'num_team' : 'team_size',
          'num_commits' : 'project_commits',
          'blau_gender' : 'gender_blau',
          'Gini_gh_ten' : 'gh_tenure_gini',
          'Core1' : 'core_dev',
          'commits' : 'user_commits',
          'propCommits' : 'commit_proportion',
          'github_tenure' : 'gh_tenure',
          'leavesNextQ' : 'leaves_next_q'},
          axis = 1,
            inplace = True)
)

In [6]:
gh_turnover.columns

Index(['project_id', 'project_domain', 'project_language', 'project_age',
       'quarter', 'team_size', 'project_commits', 'gender_blau',
       'gh_tenure_gini', 'user_id', 'core_dev', 'user_commits',
       'commit_proportion', 'gender', 'gh_tenure', 'leaves_next_q'],
      dtype='object')

In [7]:
(gh_turnover
 .agg({
     'user_commits':['mean']})
) # Calculate the mean of `user_commits`. 

user_commits
mean    339.365751

In [8]:
(gh_turnover
 .agg({
     'user_commits':['median']})
) # Calculate the median of `user_commits`.

user_commits
median          18.0

In [9]:
turnover_new = (gh_turnover
.filter(regex = '(^p|g)')
               ) # Select all columns that start with a *p* (i.e., `(^p)`) **or** contain an *g* (i.e., `(g)`). Save it as a new DataFrame named `turnover_new`.

In [10]:
turnover_new.columns # Output the columns in your notebook.

Index(['project_id', 'project_domain', 'project_language', 'project_age',
       'project_commits', 'gender_blau', 'gh_tenure_gini', 'gender',
       'gh_tenure'],
      dtype='object')

In [11]:
(turnover_new
 .query("project_commits > 9000")
) # select rows in which `project_commits` is greater than 9,000.

project_id project_domain project_language  project_age  project_commits  \
0          1101          OTHER           Python           23            22688   
1          1101          OTHER           Python           23            22688   
2          1101          OTHER           Python           23            22688   
3          1101          OTHER           Python           23            22688   
4          1101          OTHER           Python           23            22688   
..          ...            ...              ...          ...              ...   
116        1101          OTHER           Python           23            14210   
117        1101          OTHER           Python           23            14210   
118        1101          OTHER           Python           23            14210   
119        1101          OTHER           Python           23            14210   
120        1101          OTHER           Python           23            14210   

     gender_blau  gh_tenure_gini gender  gh_tenure  
0            0.0        0.422259   male        426  
1            0.0        0.422259   male         46  
2            0.0        0.422259   male        941  
3            0.0        0.422259   male       1870  
4            0.0        0.422259   male        781  
..           ...             ...    ...        ...  
116          0.0        0.411372   male        495  
117          0.0        0.411372   male        487  
118          0.0        0.411372   male        317  
119          0.0        0.411372   male        418  
120          0.0        0.411372   male        179  

[114 rows x 9 columns]

In [12]:
turnover_newer = (gh_turnover
.filter(regex = '(^p|g)')
.query("project_commits > 9000")
               ) # Using pandas piping notation, perform the previous two operations together and save it as a new DataFrame `turnover_newer`. 

In [15]:
query = (gh_turnover
.filter(regex = '(e$|s)')
.query("gh_tenure >= 1991")
) #  Using pandas piping notation, select all columns that end with the letter *e* (i.e., `e$`) 
# or contain the letter *q* (i.e., `q`). Additionally, select rows in which `gh_tenure` is greater than or equal to 1,991. 

In [16]:
(query
 .agg({
     'user_commits':['mean','median']})
) # Calculate the mean and median of `user_commits` for the new query. 

user_commits
mean             3.5
median           3.5

This mean is signficantly lower. The previous mean was 339 and the median was 18. Now both mean and median are 3.5. 

## References

### ChatPGT 
https://chatgpt.com/share/67feb8e6-fd30-8012-82ef-009a1e19f3ee
>>> I used ChatGPT this week to help me figure out how to combine regex commands. 